### DISCUSSION TITLE
activation of `hs.hotkey.modal` and `hs.chooser` are disabling sometimes my mouse event listener

### DISCUSSION THREAD

**Discussion by muescha** [Hammerspoon/hammerspoon#3369] [Q&A] [answered]
## Summary

my mouse event listener get sometimes disabled. but i don't know why. 

## What i do

i loaded all my hammerspoon stuff - and its working as expected until i fire a hotkey for an modal. at this moment my mouse listener is not working anymore.

## what i see

somehow the activation of an `hs.hotkey.modal` and a `hs.chooser` is disabling my `hs.eventtap.new({hs.eventtap.event.types.leftMouseDown},` listener.

any ideas why this happening?

## Code Snippets

a) my mouse listener defined here:
https://github.com/muescha/dot_hammerspoon/blob/master/Functions/MailMateFocus.lua

edit: code is updated - see answer

```diff
-                          hs.eventtap.new({hs.eventtap.event.types.leftMouseDown}, checkCmdClickInMailmateAndActivateMailmate):start()
+ MailMateFocus_EventTap = hs.eventtap.new({hs.eventtap.event.types.leftMouseDown}, checkCmdClickInMailmateAndActivateMailmate):start()
```

b) 2 of the modals are defined here:
https://github.com/muescha/dot_hammerspoon/blob/master/Functions/CheatSheet.lua

c) my choosers are here:
https://github.com/muescha/dot_hammerspoon/blob/master/Functions/FuzzyWindowSearch.lua

d) and here is also a chooser:
https://github.com/muescha/dot_hammerspoon/blob/f17ee62af084b236d56310106dc21d7b52a77107/Functions/CheatSheet.lua#L43-L80

## Exceptions

this modals don't disable the mouse listener:
https://github.com/muescha/dot_hammerspoon/blob/master/Functions/Experimental.lua

**Comment by muescha** [comment, marked as answer]
ok - found a solution. 

it seems it got garbage collected, because after some time the function also stopped  - so i add a global:

```lua
hs.eventtap.new()
```

changed to:

```lua
MailMateFocusEventTap = hs.eventtap.new()
```

note: a `local` variable was not working...

**Reply by muescha** [reply]
see: https://github.com/Hammerspoon/hammerspoon/issues/1743#issuecomment-385557461

> In the snippit you show above, you're not assigning the eventtap to a global variable -- as garbage collection occurs, this would cause the eventtap to be cleared automatically.
> 
> If I type in your example above into the console, it will work for a while but eventually it will stop working; or I can force it to stop immediately by typing in `collectgarbage()` into the console.

maybe add this to http://www.hammerspoon.org/faq/ ?

**Comment by asmagill** [comment]
It's been a while since I reviewed it, but I should probably get this added to the official Hammerspoon wiki at some point -- take a look at https://github.com/asmagill/hammerspoon/wiki/Variable-Scope-and-Garbage-Collection and it will show you how you can store the handler with a local variable and still prevent garbage collection.
